### PR TITLE
Natgrads

### PR DESCRIPTION
**Pull request by Daniel-Dodd**
This PR seeks to add natural gradients to GPJax, as well as two new Gaussian variational family parameterisations.

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ x] Feature
- [ ] Code style update (formatting, renaming)
- [x ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

# Current state:

Code currently only works for the natural parameterisation case. Main thing (asides from the obvious simplification of the rough codebase and improvement of the API) is that we lack natural gradients for general parameterisations.

There are two notebooks associated with this PR:
- `natgrads.ipynb` has the case the variational family is chosen as the natural parameterisation - this should work but I have not tested it since I rebased with the master branch. 
- `Natural Gradient General case.ipynb` is provided to show a rough sketch of what the general case might look like, that will involve the user defining a bijection between their parameterisation and the natural parameterisation.

There are some unit tests provided for the new variational families, and for `natural_gradients.py` in its current state, as well as changes made in `parameters.py`. 

As a final note, it is likely that some of the functions defined in `natural_gradients.py` for stopping gradients might be better generalised and added to `abstractions.py` or `parameters.py`

### CODE REVIEW

**Comment by Daniel-Dodd** [comment]
@thomaspinder it would be good to get your thoughts on the code so far (some of it rather rough). I believe the implementation works. The main issue is that the notebook needs writing before merging, I'll get on with this in the meantime. I'll need to test some benchmarks against `fit_batches` once #99 is merged, to see if performance is up to scratch.

File: gpjax/natural_gradients.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+from .variational_inference import StochasticVI
+
+DEFAULT_JITTER = get_defaults()["jitter"]
+
+
+def natural_to_expectation(
+    natural_moments: dict, jitter: float = DEFAULT_JITTER
+) -> dict:
+    """
+    Converts natural parameters to expectation parameters.
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:38]: Can we give the equations or at least a reference to the Salimbeni paper here?

File: gpjax/natural_gradients.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            return xi_elbo(params, batch)
+
+        value, dL_dhyper = value_and_grad(loss_fn)(params, batch)
+
+        return value, dL_dhyper
+
+    return nat_grads_fn, hyper_grads_fn
+
+
+def fit_natgrads(
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:240]: We need a unit test(s) for this fn.

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Started a review but will finish more tomorrow.

**Comment by Daniel-Dodd** [comment]
@thomaspinder I have rebased the branch with master and would appreciate a review.

On my end, I need to:
- Update and add some tests
- Write the notebook

File: docs/nbs/natgrads.ipynb [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    "import matplotlib.pyplot as plt\n",
+    "from jax import jit, lax\n",
+    "import optax as ox\n",
+    "\n",
+    "import gpjax as gpx\n",
+    "from gpjax.abstractions import progress_bar_scan\n",
+    "\n",
+    "#Set seed for reproducibility:\n",
+    "import tensorflow as tf\n",
+    "tf.random.set_seed(4)\n",
```
Comment (thomaspinder) [inline docs/nbs/natgrads.ipynb:41]: Redundant?

**Review by thomaspinder — CHANGES_REQUESTED** [review]

File: gpjax/natural_gradients.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    ExpectationVariationalGaussian,
+    NaturalVariationalGaussian,
+)
+from .variational_inference import StochasticVI
+
+DEFAULT_JITTER = get_defaults()["jitter"]
+
+
+def natural_to_expectation(
+    natural_moments: dict, jitter: float = DEFAULT_JITTER
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:35]: Import `Dict` from typing in your imports and then use `Dict` as the type.

File: gpjax/natural_gradients.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+) -> dict:
+    """
+    Converts natural parameters to expectation parameters.
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:38]: Can we add a little more mathematical context here on the operation that is being applied. Even if it's as simple as giving the equations. I know there in the inline comments, but this is what will be rendered in the docs.

File: gpjax/natural_gradients.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    variational_family: AbstractVariationalFamily,
+    train_data: Dataset,
+) -> tp.Callable[[dict, Dataset], float]:
+    """
+    Construct evidence lower bound (ELBO) for varational Gaussian under the expectation parameterisation.
+    Args:
+        posterior: An instance of AbstractPosterior.
+        variational_family: An instance of AbstractVariationalFamily.
+    Returns:
+        Callable: A function that computes ELBO.
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:89]: typo - the ELBO.

File: gpjax/natural_gradients.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    }
+
+
+def _expectation_elbo(
+    posterior: AbstractPosterior,
+    variational_family: AbstractVariationalFamily,
+    train_data: Dataset,
+) -> tp.Callable[[dict, Dataset], float]:
+    """
+    Construct evidence lower bound (ELBO) for varational Gaussian under the expectation parameterisation.
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:84]: typo - variational

File: gpjax/natural_gradients.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+    Args:
+        posterior: An instance of AbstractPosterior.
+        variational_family: An instance of AbstractVariationalFamily.
+    Returns:
+        Callable: A function that computes ELBO.
+    """
+    evg = ExpectationVariationalGaussian(
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:91]: avoid variable names this short/ambiguous.

File: gpjax/natural_gradients.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    Args:
+        params: A dictionary of parameters.
+    Returns:
+        tp.Dict: A dictionary of parameters with stopped gradients.
+    """
+    trainables = build_trainables_true(params)
+    moment_trainables = build_trainables_false(params["variational_family"]["moments"])
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:128]: could these two functions be combined and simplified. Something like `stop_gradients` with a boolean argument for natural parameters or moments? Seems excessive right now.

File: gpjax/natural_gradients.py [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+
+def natural_gradients(
+    stochastic_vi: StochasticVI,
+    train_data: Dataset,
+    transformations: dict,
+    # bijector = tp.Optional[dx.Bijector] = Identity, #bijector: A bijector to convert between the user chosen parameterisation and the natural parameters.
+) -> tp.Tuple[tp.Callable[[dict, Dataset], dict]]:
+    """
+    Computes natural gradients for variational Gaussian.
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:138]: Be explicit about what the gradient is being computed with respect to i.e., the natural parameters.

File: gpjax/natural_gradients.py [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
+
+
+def natural_gradients(
+    stochastic_vi: StochasticVI,
+    train_data: Dataset,
+    transformations: dict,
+    # bijector = tp.Optional[dx.Bijector] = Identity, #bijector: A bijector to convert between the user chosen parameterisation and the natural parameters.
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:135]: Can this be deleted?

File: gpjax/natural_gradients.py [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+        tp.Dict: A dictionary of parameters with stopped gradients.
+    """
+    trainables = build_trainables_true(params)
+    moment_trainables = build_trainables_false(params["variational_family"]["moments"])
+    trainables["variational_family"]["moments"] = moment_trainables
+    params = trainable_params(params, trainables)
+    return params
+
+
+def natural_gradients(
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:131]: Is this fn. better suited to be in the abstractions file?

File: gpjax/natural_gradients.py [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+            return xi_elbo(params, batch)
+
+        value, dL_dhyper = value_and_grad(loss_fn)(params, batch)
+
+        return value, dL_dhyper
+
+    return nat_grads_fn, hyper_grads_fn
+
+
+def fit_natgrads(
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:240]: As above - should this be moved into abstractions?

File: gpjax/natural_gradients.py [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+            # Stop gradients for the moment parameters.
+            params = _stop_gradients_moments(params)
+
+            return xi_elbo(params, batch)
+
+        value, dL_dhyper = value_and_grad(loss_fn)(params, batch)
+
+        return value, dL_dhyper
+
+    return nat_grads_fn, hyper_grads_fn
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:237]: Fine for now, but this function is __very__ long (~100 lines). It'd be good if we could break it up in the future. If it's easy, then I'd suggest doing it within this PR. If not, please raise an issue to refactor this fn.

File: gpjax/natural_gradients.py [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+    transformations: tp.Dict,
+    train_data: Dataset,
+    batch_size: int,
+    moment_optim,
+    hyper_optim,
+    key: PRNGKeyType,
+    n_iters: tp.Optional[int] = 100,
+    log_rate: tp.Optional[int] = 10,
+) -> tp.Dict:
+
```
Comment (thomaspinder) [inline gpjax/natural_gradients.py:253]: We should have a docstring here explaining why we optimise the variational distribution then the hyperparameters i.e., stale posteriors. Also happy for this information to be duplicated in the notebook.

File: gpjax/parameters.py [hunk 16]
```diff
@@ -188,6 +187,20 @@ def transform(params: tp.Dict, transform_map: tp.Dict) -> tp.Dict:
...
+    Args:
+        params (tp.Dict): The parameter set for which trainable statuses should be derived from.
+
+    Returns:
+        tp.Dict: A dictionary of identity forward/backward bijectors. The dictionary is equal in structure to the input params dictionary.
+    """
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+
+    return jax.tree_map(lambda _: Identity.forward, prior_container)
```
Comment (thomaspinder) [inline gpjax/parameters.py:201]: jax.tree_util.tree_map

File: gpjax/parameters.py [hunk 17]
```diff
@@ -278,6 +291,22 @@ def build_trainables(params: tp.Dict) -> tp.Dict:
...
+    Args:
+        params (tp.Dict): The parameter set for which trainable statuses should be derived from.
+
+    Returns:
+        tp.Dict: A dictionary of boolean trainability statuses. The dictionary is equal in structure to the input params dictionary.
+    """
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+    # Set all values to zero
+    prior_container = jax.tree_map(lambda _: False, prior_container)
```
Comment (thomaspinder) [inline gpjax/parameters.py:306]: tree_util

File: gpjax/variational_families.py [hunk 18]
```diff
@@ -45,21 +45,27 @@ def predict(self, *args: Any, **kwargs: Any) -> dx.Distribution:
...
 
     """
 
-    prior: Prior
-    inducing_inputs: f64["M D"]
-    name: str = "Variational Gaussian"
-    variational_mean: Optional[f64["M Q"]] = None
-    variational_root_covariance: Optional[f64["M M"]] = None
+    variational_mean: Optional[f64["N D"]] = None
+    variational_root_covariance: Optional[f64["N D"]] = None
```
Comment (thomaspinder) [inline gpjax/variational_families.py:67]: Not `f64["N N"]`?

File: gpjax/variational_families.py [hunk 19]
```diff
@@ -259,6 +267,282 @@ def predict_fn(test_inputs: f64["N D"]) -> dx.Distribution:
         return predict_fn
 
 
+@dataclass
+class NaturalVariationalGaussian(AbstractVariationalGaussian):
```
Comment (thomaspinder) [inline gpjax/variational_families.py:271]: This object is wonderfully structured and coded.

**Review by thomaspinder — CHANGES_REQUESTED** [review]
Several comments here, many of them suggestions, not hard requires. A general comment is that more information is needed in the docstrings - when the docs are built, this is what'll be rendered. I'm very happy for the maths within docstrings to then be moved into a notebook.

File: gpjax/natural_gradients.py (hunk 4 above)
Comment (Daniel-Dodd) [reply gpjax/natural_gradients.py:35]: Absolutely. This is something that is inconsistent across the package. Will address this in the next commit.

File: gpjax/natural_gradients.py (hunk 5 above)
Comment (Daniel-Dodd) [reply gpjax/natural_gradients.py:38]: Good suggestion.

File: gpjax/natural_gradients.py (hunk 9 above)
Comment (Daniel-Dodd) [reply gpjax/natural_gradients.py:128]: Good suggestion. Will look into this.

File: gpjax/natural_gradients.py (hunk 13 above)
Comment (Daniel-Dodd) [reply gpjax/natural_gradients.py:240]: I think that would be wise - was wondering the same.

File: gpjax/variational_families.py (hunk 18 above)
Comment (Daniel-Dodd) [reply gpjax/variational_families.py]: Variational families are revised in d84ede9.

**Comment by Daniel-Dodd** [comment]
@thomaspinder for the commits d84ede9, e0b60d4 and bd6e4aa, I have addressed many of your comments and suggestions.

Outstanding issues are to write unit tests, the notebook, and explanation of parameter optimisation order.

**Comment by Daniel-Dodd** [comment]
Hi @thomaspinder, I believe I have addressed your comments so far. I expect all functions to be covered by unit tests now (but we'll see what CodeCov says). I would appreciate a review on the latest code changes - and I expect we can improve it further.

One key thing that I dislike, is that the `fit_natgrads` abstraction takes it fist argument as the variational inference strategy, while `fit` and `fit_batches` take in objectives. I'm not sure if you had any thoughts on this?


I am going to start work on writing the notebook today.

File: gpjax/abstractions.py [hunk 20]
```diff
@@ -192,6 +194,80 @@ def step(carry, iter_num__and__key):
...
+    Returns:
+        InferenceState: A dataclass comprising optimised parameters and training history.
+    """
+
+    hyper_state = hyper_optim.init(params)
+    moment_state = moment_optim.init(params)
+
+    nat_grads_fn, hyper_grads_fn = natural_gradients(
+        stochastic_vi, train_data, transformations
+    )
```
Comment (thomaspinder) [inline gpjax/abstractions.py:233]: These lines seem to be the only ones that require the `stochastic_vi` object. Can we simply move this outside of the training loop and then just make the function accept two objective_fn; one for the hyperparams and one for the natgrads?

File: gpjax/abstractions.py [hunk 21]
```diff
@@ -192,6 +194,80 @@ def step(carry, iter_num__and__key):
...
+
+    # Initial natural gradient step to improve bound for hyperparameters:
+    batch = get_batch(train_data, batch_size, key)
+    loss_val, loss_gradient = nat_grads_fn(params, trainables, batch)
+    updates, moment_state = moment_optim.update(loss_gradient, moment_state, params)
+    params = optax.apply_updates(params, updates)
```
Comment (thomaspinder) [inline gpjax/abstractions.py:239]: Do we want to run this for several iterations? We could perhaps add an optional argument similar to the number of warmup steps in a HMC sampler? Something like `natgrad_warmup_nits: int = 50`?

File: gpjax/natural_gradients.py (hunk 12 above)
Comment (thomaspinder) [reply gpjax/natural_gradients.py:143]: So are we moving this or not?

File: gpjax/parameters.py [hunk 22]
```diff
@@ -173,21 +172,35 @@ def inverse(bijector):
...
+
+    Returns:
+        Dict: A dictionary of identity forward/backward bijectors. The dictionary is equal in structure to the input params dictionary.
+    """
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+
+    return jax.tree_util.tree_map(lambda _: Identity.forward, prior_container)
+
+
```
Comment (thomaspinder) [inline gpjax/parameters.py:203]: I don't get why this is needed?

File: gpjax/parameters.py [hunk 23]
```diff
@@ -278,12 +291,28 @@ def build_trainables(params: tp.Dict) -> tp.Dict:
...
+        params (Dict): The parameter set for which trainable statuses should be derived from.
+
+    Returns:
+        Dict: A dictionary of boolean trainability statuses. The dictionary is equal in structure to the input params dictionary.
+    """
+    # Copy dictionary structure
+    prior_container = deepcopy(params)
+    # Set all values to zero
+    prior_container = jax.tree_util.tree_map(lambda _: False, prior_container)
+    return prior_container
```
Comment (thomaspinder) [inline gpjax/parameters.py:307]: I still don't like that this is two functions - it it feels like an argument would resolve it:
```python
def build_trainables_true(params: Dict, status: bool = False) -> Dict:
    """Construct a dictionary of trainable statuses for each parameter. By default, every parameter within the model is trainable.
    Args:
        params (Dict): The parameter set for which trainable statuses should be derived from.
        status (bool): Whether or not gradients should be taken with respect to this parameter.
    Returns:
        Dict: A dictionary of boolean trainability statuses. The dictionary is equal in structure to the input params dictionary.
    """
    # Copy dictionary structure
    prior_container = deepcopy(params)
    # Set all values to zero
    prior_container = jax.tree_util.tree_map(lambda _: status, prior_container)
    return prior_container

**Review by thomaspinder — CHANGES_REQUESTED** [review]
A few small comments from me. When these are resolved and we're happy with the notebook, I think it's ready to merge.

File: gpjax/parameters.py (hunk 22 above)
Comment (Daniel-Dodd) [reply gpjax/parameters.py:203]: In an upcoming commit, I have changed `gpx.transform` to accept `transformation_map = None` to give an identity transformation.

File: gpjax/parameters.py (hunk 23 above)
Comment (Daniel-Dodd) [reply gpjax/parameters.py:283]: Resolved in cb45fbf.

File: gpjax/parameters.py (hunk 22 above)
Comment (Daniel-Dodd) [reply gpjax/parameters.py]: Resolved in https://github.com/thomaspinder/GPJax/commit/cb45fbfc3ffa449cb216c11250edb0fe4f20051b.

File: gpjax/abstractions.py (hunk 21 above)
Comment (thomaspinder) [reply gpjax/abstractions.py:239]: I'm fine with that.
